README.md
# Perpetual Calendar based on rules.

- Under construcion


ChangeLog.md
# Changelog for perpetual

## Unreleased changes


tests/tests.hs

{-# LANGUAGE OverloadedStrings #-}

module Test.USARules where

import qualified Data.Text as T
import Perpetual.DateTerm
import Perpetual.ScheduleTerm
import Perpetual.ScheduleParser 

getRules file = do Right (RulesEnv v rs) <- parseFile file
                   return (RulesEnv v (mergeSort rs))

usaRules = getRules "Rules/usRules.txt"

churchRules = getRules "Rules/churchRules.txt"

usEx dt = matchRules <$> usaRules <*> return dt
        
churchEx dt = matchRules <$> churchRules <*> return dt


tests/USARules.hs
{-# LANGUAGE OverloadedStrings #-}

module Test.USARules where

import qualified Data.Text as T
import Perpetual.DateTerm
import Perpetual.ScheduleTerm
import Perpetual.ScheduleParser 

rulesUsaTxt :: [T.Text]
rulesUsaTxt = [
  "Holiday \"New Year's Day\" January 1, *",
  "Holiday \"Birthday of Martin Luther King\" 3rd Monday after January 1, *",
  "Holiday \"Inauguration Day\" every 4 years since January 20, 1937",
  "Holiday \"Washington's Birthday\" 3rd Monday after February 1, *",
  "Holiday \"Armed Forces Day\" 3rd Sunday after May 1, *",
  "Holiday \"Memorial Day\" last Monday in May *",
  "Holiday \"Flag Day\" June 14, *",
  "Holiday \"United States of America's Independence Day\" July 4, *",
  "Holiday \"Labor Day\" first Monday after September 1, *",
  "Holiday \"Columbus Day\" 2nd Monday after October 1, *",
  "Holiday \"Election Day\" Tuesday after November 2, *",
  "Holiday \"Veterans Day\" Monday after November 11, *",
  "Holiday \"Thanksgiving Day\" 4th Thursday after November 1, *",
  "Remark \"Black Friday\" 4th Friday after November 1, *",
  "Holiday \"Christmas Day\" December 25, *",
  "Holyday \"Easter\" Easter *",
  "Remark \"Monday remark\" Monday after every 7 days since January 1, *",
  "Remark \"Last month day\" last day of * *",
  "Non-business \"Vacations\" {Monday .. Sunday} from December 23, * to December 31, *",
  "Non-business \"Vacations\" {Monday .. Sunday} from January 1, * to January 3, *",
  "Non-business \"Weekend\" Saturday after *",
  "Non-business \"Weekend\" Sunday after *",
  "Non-business \"Holy Friday\" Friday before Easter *",
  "Special schedule \"Last Friday\" [8:00 - 12:00] last Friday in * *",
  "Special schedule \"Easter days\" [8:00 - 12:00] {Monday .. Friday} from Thursday before Easter * to Wednesday after Easter *",
  "Business schedule [8:00 - 12:30pm; 2:00pm - 6:00pm] {Monday .. Friday} from January 1, * to December 31, *"
  ]

rulesUsaP = map (parseOnly parseCalendarTerm) rulesUsaTxt
rulesUsa = mergeSort . map (either (error . show) id) $ rulesUsaP




src/Perpetual/DateTerm.hs

module Perpetual.DateTerm where

import Data.Time.Calendar (fromGregorian,
                           toGregorian,
                           gregorianMonthLength,
                           addGregorianMonthsClip,
                           addDays,diffDays, Day)
import Data.Time.Calendar.Easter (gregorianEaster)
import Data.Time.Calendar.WeekDate (toWeekDate)
import Control.Applicative ((<$>),(<*>),pure)
import qualified Data.Map as M 
import qualified Data.Text as T

data Month = January
           | February
           | March
           | April
           | May
           | June
           | July
           | August
           | September
           | October
           | November
           | December
           deriving (Show,Eq,Ord,Enum)

data MonthTerm = MonthCst Month
               | MonthVar Identifier
               | MonthOf DatePattern
               deriving (Show,Eq,Ord)

data WeekDay = Monday
             | Tuesday
             | Wednesday
             | Thursday
             | Friday
             | Saturday
             | Sunday
             deriving (Show,Eq,Ord,Enum)  

data WeekDayTerm = WeekDayCst WeekDay               
                 | WeekDayOf DatePattern
                 deriving (Show,Eq,Ord)


type MonthDay = Int

data MonthDayTerm = MonthDayCst MonthDay
                  | MonthDayVar Identifier
                  | MonthDayOf DatePattern
                  deriving (Show,Eq,Ord)

type Year = Integer

type Date = (Integer,Int,Int)

data YearTerm = YearCst Year
              | YearVar Identifier
              | YearOf DatePattern
              deriving (Show,Eq,Ord)

type Identifier = T.Text

data DatePattern = DatePattern YearTerm MonthTerm MonthDayTerm  
                 | Easter YearTerm 
                 | DateVar Identifier  
                 | DateDaysAfter DatePattern Integer
                 | DateDaysBefore DatePattern Integer  
                 | DateMonthFirstWeekDay YearTerm MonthTerm WeekDayTerm
                 | DateMonthLastWeekDay YearTerm MonthTerm WeekDayTerm
                 | DateFirstDay YearTerm MonthTerm
                 | DateLastDay YearTerm MonthTerm
                 | DateWeekDayAfter DatePattern WeekDayTerm -- 
                 | DateWeekDayBefore DatePattern WeekDayTerm -- 
                 | DateNthWeekDay DatePattern WeekDayTerm Integer  
                 | DateNYearsSince DatePattern Integer
                 | DateNMonthsSince DatePattern Integer
                 | DateNDaysSince DatePattern Integer  
                 deriving (Show,Eq,Ord)

data DateValue = DateValue {
                     weekDayV :: WeekDay,
                     monthDayV :: MonthDay,
                     monthV :: Month,
                     yearV :: Year,
                     weekNumberV :: Int,
                     dateV :: Date
                 }
                 deriving (Show,Eq,Ord)

type DateContext = M.Map T.Text DateValue

year  (y,_,_) = y
month (_,m,_) = m
day   (_,_,d) = d

weekDay :: Int -> WeekDay
weekDay n = toEnum (n - 1)

fromGregorian' :: (Integer,Int,Int) -> Day
fromGregorian' (y,m,d) = fromGregorian y m d

getDateVar :: T.Text -> DateContext -> Either String DateValue
getDateVar k ctx = maybe (Left $ "undefined date variable '"++ T.unpack k ++"'") Right $ k `M.lookup` ctx

defaultDateIdentifier :: T.Text
defaultDateIdentifier = T.pack "_default"

getDefaultDateValue :: DateContext -> Either String DateValue
getDefaultDateValue = getDateVar defaultDateIdentifier 

evalWeekDayTerm :: WeekDayTerm -> DateContext -> Either String Int
evalWeekDayTerm (WeekDayCst wd) _ = pure $ fromEnum wd + 1
evalWeekDayTerm (WeekDayOf t) ctx = day . toWeekDate . fromGregorian' <$> evalDatePattern t ctx 

evalMonthDayTerm :: MonthDayTerm -> DateContext -> Either String MonthDay
evalMonthDayTerm (MonthDayCst d) _ = pure d
evalMonthDayTerm (MonthDayOf t) ctx = day <$> evalDatePattern t ctx
evalMonthDayTerm (MonthDayVar ident) ctx = monthDayV <$> getDateVar ident ctx

evalMonthTerm :: MonthTerm -> DateContext -> Either String Int
evalMonthTerm (MonthCst m) _ = pure $ fromEnum m + 1 
evalMonthTerm (MonthOf t) ctx = month <$> evalDatePattern t ctx
evalMonthTerm (MonthVar ident) ctx = (+1) . fromEnum . monthV <$> getDateVar ident ctx

evalYearTerm :: YearTerm -> DateContext -> Either String Year
evalYearTerm (YearCst n) _ = return n
evalYearTerm (YearOf t) cxt = year <$> evalDatePattern t cxt
evalYearTerm (YearVar ident) ctx = yearV <$> getDateVar ident ctx

normWeekDay :: Integral n => n -> n
normWeekDay n = (7 + n) `mod` 7

evalDatePattern :: DatePattern -> DateContext -> Either String Date
evalDatePattern (DatePattern yt mt dt) ctx = 
  (,,) <$> evalYearTerm yt ctx <*> evalMonthTerm mt ctx <*> evalMonthDayTerm dt ctx
evalDatePattern (Easter yt) ctx = 
  toGregorian . gregorianEaster <$> evalYearTerm yt ctx
evalDatePattern (DateDaysAfter dt n) ctx = 
  toGregorian . addDays n . fromGregorian' <$>  evalDatePattern dt ctx
evalDatePattern (DateDaysBefore dt n) ctx = 
  toGregorian . addDays (-n) . fromGregorian' <$> evalDatePattern dt ctx
evalDatePattern (DateMonthFirstWeekDay yt mt wdt) ctx =
  evalDatePattern (DateNthWeekDay (DatePattern yt mt (MonthDayCst 1)) wdt 1) ctx
evalDatePattern (DateMonthLastWeekDay yt mt wdt) ctx =
   lastWeekDay <$> evalYearTerm yt ctx <*> evalMonthTerm mt ctx <*> evalWeekDayTerm wdt ctx
  where lastWeekDay y m wd' = (y,m,ld - normWeekDay (wd - wd'))
             where ld = gregorianMonthLength y m
                   d  = fromGregorian y m ld
                   wd = day . toWeekDate $ d 
evalDatePattern (DateFirstDay yt mt) ctx =
  (,,) <$> evalYearTerm yt ctx <*> evalMonthTerm mt ctx <*> return 1
evalDatePattern (DateLastDay yt mt) ctx = do
  (\y m -> (y,m,gregorianMonthLength y m)) <$> evalYearTerm yt ctx <*> evalMonthTerm mt ctx 
evalDatePattern (DateWeekDayAfter dt wdt) ctx = evalDatePattern (DateNthWeekDay dt wdt 1) ctx 
evalDatePattern (DateWeekDayBefore dt wdt) ctx = do
     weekDayUpto <$> evalDatePattern dt ctx <*> evalWeekDayTerm wdt ctx
  where weekDayUpto (y,m,dd) wd = (y,m,dd - normWeekDay (d - wd))
            where d = day . toWeekDate $ fromGregorian y m dd
evalDatePattern (DateNthWeekDay dt wdt n) ctx = do
    nthWeekDay <$> evalDatePattern dt ctx <*> evalWeekDayTerm wdt ctx
  where nthWeekDay (y,m,dd) wd = toGregorian gd'
            where gd = fromGregorian y m dd
                  d = day . toWeekDate $ gd
                  gd' = addDays ((normWeekDay . toInteger $ wd - d) + 7*(n - 1)) gd
evalDatePattern (DateVar ident) ctx = dateV <$> getDateVar ident ctx
evalDatePattern (DateNYearsSince dt n) ctx = do 
        (y,m,d) <- evalDatePattern dt ctx
        (y',_,_) <- dateV <$> getDefaultDateValue ctx
        let y'' = if y <= y' then y' - ((y' - y) `mod` n) else y 
        return (y'',m,d)
evalDatePattern (DateNMonthsSince dt n) ctx = do
        date@(y,m,d) <- evalDatePattern dt ctx
        date'@(y',m',d') <- dateV <$> getDefaultDateValue ctx
        let deltaM = toInteger $ diffMonths date' date
            sinceD = fromGregorian' date
            date'' = toGregorian $ addGregorianMonthsClip (deltaM - (deltaM `mod` n)) sinceD
        return $ if date > date' then date else date''
evalDatePattern (DateNDaysSince dt n) ctx = do
  date@(y,m,d) <- evalDatePattern dt ctx
  date'@(y',m',d') <- dateV <$> getDefaultDateValue ctx
  let days  = fromGregorian' date
      days' = fromGregorian' date'
      deltaD = toInteger $ diffDays days' days
      date'' = toGregorian $ addDays (deltaD - (deltaD `mod` n)) days
  return $ if date > date' then date else date''

diffMonths :: Date -> Date -> Integer
diffMonths d1@(y,m,_) d2@(y',m',_) | d1 <= d2 && y == y' = toInteger (m' - m)
diffMonths d1@(y,m,_) d2@(y',m',_) | d1 < d2 = 13 - toInteger m + (y' - y - 1)*12 + toInteger (m' - 1)
diffMonths d1@(y,m,_) d2@(y',m',_) = -(diffMonths d2 d1) 

matchWithDate :: DatePattern -> Date -> DateContext -> Bool
matchWithDate dt date ctx = either (const False) id $ (== date) <$> evalDatePattern dt ctx     


getDateValueFromDay :: Day -> DateValue
getDateValueFromDay day = DateValue { monthDayV = monthDay,
                                      weekDayV = weekDay,
                                      monthV = month,
                                      yearV = year,
                                      weekNumberV = weekN,
                                      dateV = date } 
                        where date@(year,monthN,monthDay) = toGregorian day
                              month = toEnum (monthN - 1)
                              (_,weekN,weekDayN) = toWeekDate day
                              weekDay = toEnum (weekDayN - 1)

getDateValueFromDate :: Date -> DateValue
getDateValueFromDate date@(year,monthN,monthDay) =
                          DateValue { monthDayV = monthDay,
                                      weekDayV = weekDay,
                                      monthV = month,
                                      yearV = year,
                                      weekNumberV = weekN,
                                      dateV = date }
                   where month = toEnum (monthN - 1)
                         (_,weekN,weekDayN) = toWeekDate . fromGregorian' $ date
                         weekDay = toEnum (weekDayN - 1)

defaultDateContext dt = M.singleton defaultDateIdentifier (getDateValueFromDate dt)


src/Perpetual/ScheduleTerm.hs
module Perpetual.ScheduleTerm where

import Perpetual.DateTerm
import Data.Time.LocalTime
import Data.Time.Calendar (fromGregorian,toGregorian)
import Data.Time.Calendar.WeekDate (toWeekDate)
import qualified Data.Text as T
import qualified Data.Map as M
import Control.Applicative ((<$>),(<*>))

type Duration = TimeOfDay

type DaySchedule = [IntervalTime] 

type Description = T.Text

data IntervalTime = IntervalTime { beginTime :: TimeOfDay, endTime :: TimeOfDay , intervalDescription :: Description }
                  deriving (Show,Eq,Ord)

data InfoType = Remark -- ^ only for information 
              | Holiday -- ^ The day is a Holiday
              | Description -- ^ Description of a rule  
              deriving (Show,Eq,Enum,Ord)

data TaskType = FixedTask IntervalTime
              | FlexibleTask Duration
              deriving (Show,Eq,Ord)

data WorkableType = NonBusiness -- ^ The day is a non business Day
                  | Special { daySchedule :: DaySchedule } -- ^ The day is a business Day with special schedule
                  | Business { daySchedule :: DaySchedule } -- ^ The day is a Normal labour day
                  deriving (Show,Eq,Ord)

data RuleType = InfoRule InfoType
              | TaskRule TaskType 
              | WorkableRule WorkableType Bool -- ^ Apply to holiday if True
              deriving (Show,Eq,Ord)

data CalendarRule = DayRule  { ruleType :: RuleType,
                               description:: Description,
                               datePattern :: [DatePattern] } -- ^ rules applied  
                  | RangeRule { ruleType :: RuleType,
                                description :: Description,
                                sinceDate :: DatePattern,
                                untilDate ::  DatePattern,
                                weekDays :: [WeekDay] } -- ^ range rule
                  deriving (Show,Eq)  

data CalendarValue = CalendarValue {
                         dt :: DateValue,
                         wt  :: Maybe WorkableType,
                         isHoliday :: Bool,
                         tasks :: [(TaskType,Description)],
                         info  :: [(InfoType,Description)]
                         }
                   deriving (Show,Eq)  

instance Ord CalendarRule where
  compare ct ct' = compare (ruleType ct) (ruleType ct')

data RulesEnv = RulesEnv { variables :: [(T.Text,DatePattern)] , rules :: [CalendarRule]  }
              deriving (Show)  

wellDefinedInterval :: IntervalTime -> Bool
wellDefinedInterval (IntervalTime b e _) = b < e

notOverlapping :: IntervalTime -> IntervalTime -> Bool
notOverlapping (IntervalTime b e _) (IntervalTime b' e' _) = e < b' || e' < b

wellDefinedSchedule :: [IntervalTime] -> Bool
wellDefinedSchedule [] = True
wellDefinedSchedule (x:xs) = wellDefinedInterval x && all (notOverlapping x) xs
                             && wellDefinedSchedule xs 

matchCalendarSchedule :: CalendarRule -> Date -> DateContext -> Bool
matchCalendarSchedule (DayRule _ _ dps) date ctx = or $ map (\dp -> matchWithDate dp date ctx) dps
matchCalendarSchedule (RangeRule _ _ s u wds) date ctx = matchRange s u date ctx && wd `elem` wds
     where wd = toEnum . (\x -> x - 1) . day . toWeekDate . fromGregorian' $ date
           matchRange s u date ctx = either (const False) id ev
           ev = do sd <- evalDatePattern s ctx
                   ud <- evalDatePattern u ctx
                   return $ sd <= date && date <= ud

merge xs [] = xs
merge [] ys = ys
merge (x:xs) (y:ys) = if x <= y
                      then (x:merge xs (y:ys))
                      else (y:merge (x:xs) ys)

mergeSort [] = []
mergeSort [x] = [x]
mergeSort xs = merge (mergeSort ys) (mergeSort ys')
             where (ys,ys') = splitAt (length xs `div` 2) xs
                   
-- Precondition: The rules must be sorted according the day type
matchRules :: RulesEnv -> Date -> CalendarValue
matchRules (RulesEnv env rules) dt0 = matchRules' rules (CalendarValue dt' Nothing False [] [])
  where dt = (\(y,m,d) -> toGregorian . fromGregorian y m $ d) dt0
        matchRules' [] cv = cv
        matchRules' (r:rs) cv | matchCalendarSchedule r dt ctx =
          case ruleType r of
            (InfoRule infoT) -> matchRules' rs (cv { isHoliday = isHoliday cv || infoT == Holiday,
                                                     info = (infoT,description r):info cv})
            (TaskRule taskT) -> matchRules' rs (cv {tasks = (taskT,description r):tasks cv}) 
            (WorkableRule wtype h) -> if h == isHoliday cv
                                      then cv { wt = (Just wtype), info = inf } 
                                      else matchRules' rs cv
                                where inf = if d == T.empty then info cv else (Description,d):info cv
                                      d = description r
        matchRules' (_:rs) cv = matchRules' rs cv
        ctx0 = defaultDateContext dt
        Right dt' = getDateVar defaultDateIdentifier ctx0
        ctx = foldr (\(v,dp) m -> either (const m) (flip (M.insert v) m . getDateValueFromDate) $ evalDatePattern dp m) ctx0 env 


smatchRules :: RulesEnv -> Date -> CalendarValue
smatchRules (RulesEnv env rules) dt0 = matchRules (RulesEnv env $ mergeSort rules) dt0


src/Perpetual/ScheduleParser.hs
{-# LANGUAGE OverloadedStrings, TupleSections #-}

module Perpetual.ScheduleParser where

{-
import Data.Attoparsec.Text as P
import Data.Attoparsec.Combinator 
import Control.Applicative
-}
import Control.Applicative ((<|>),(<$),(<*>),(<$>),(<*),(*>))
-- Parsec
import Text.Parsec.Prim (many,try,lookAhead,Parsec(..),updateState,getState,setState,runParser)
import Text.Parsec.Combinator (many1)
import Text.Parsec.Char (spaces,digit,lower,char,alphaNum,satisfy,string)
import Text.Parsec.Text hiding (Parser)
import Text.Parsec.Error
-- *Parsec
import Data.List (nub)

import Data.Time.LocalTime

import Data.Char

import Perpetual.DateTerm
import Perpetual.ScheduleTerm

import qualified Data.Text as T
import qualified Data.Text.IO as T


type Parser = Parsec T.Text RulesEnv

-- Parsec
-- type Parser = Parsec T.Text ()
-- *Parsec

parseEnum :: (Enum a,Show a) => Parser a
parseEnum = cases
          <|> do let e = toEnum 0
                 fail ("No enum case ["++ show e ++"..]")
                 return e
          where cases = foldl1 (<|>) $ map
                               (\d -> d <$ (try . accept . T.pack . show $ d))
                               [toEnum 0 .. ]

parseWeekDay :: Parser WeekDay
parseWeekDay = parseEnum

parseMonth :: Parser Month
parseMonth = parseEnum

parseDescription :: Parser T.Text
parseDescription = myToken (char '\"' *> (T.pack <$> many (satisfy (/= '\"'))) <*char '\"')
                 <|> return T.empty
{-
parseDescription = char '\"' *> (P.takeWhile (/= '\"')) <*char '\"'
-}

parseYear :: Parser Year
parseYear = satM (> 1858) integer
          <|> fail "The year must be at least 1859"

parseOrdinal :: Parser Integer
parseOrdinal = try (
             1  <$ (accept "first" <|> accept "1st")
             <|> 2  <$ (accept "second" <|> accept "2nd")
             <|> 3  <$ (accept "third" <|> accept "3rd")
             <|> 4  <$ accept "fourth"
             <|> 5  <$ accept "fifth"
             <|> 6  <$ accept "sixth"
             <|> 7  <$ accept "seventh"
             <|> 8  <$ accept "eighth"
             <|> 9  <$ accept "ninth"
             <|> 10 <$ accept "tenth"
             <|> satM (>3) integer <* accept "th")

satM :: (Monad m,Show b) => (b -> Bool) -> m b -> m b
satM cnd parser = do v <- parser
                     if cnd v
                     then return v
                     else fail ("Unsatisfied condition with value '" ++ show v ++ "'")

parseVar :: Parser T.Text
parseVar =   defaultDateIdentifier <$ accept "*"
         <|> do v <- parseIdentifier
                s <- getState
                case lookup v (variables s) of
                  Nothing -> fail $ "undefined variable '"++ T.unpack v ++"'"
                  (Just _) -> return v

parseIdentifier :: Parser T.Text
parseIdentifier = T.pack <$ spaces <*> (char '$' *> ((:) <$> lower <*> many alphaNum)) <* spaces

parseDatePatternDefList :: Parser [DatePattern]
parseDatePatternDefList = (:) <$> parseDateDef <*> many (accept ";" *> parseDateDef)

parseDateDef :: Parser DatePattern
parseDateDef = do var@(v,t) <- try ((,) <$> parseIdentifier <* accept "=" <*> parseDatePattern)
                  s <- getState
                  case lookup v (variables s) of
                    Nothing -> do setState (s { variables = var : variables s })
                                  return t
                    (Just _) -> fail $ "variable '"++ T.unpack v ++"' already defined"
             <|> parseDatePattern

ignore :: Parser a -> Parser ()
ignore = (() <$)

parseWeekDayTerm :: Parser WeekDayTerm
parseWeekDayTerm = WeekDayOf <$ accept "week" <* accept "day" <* accept "of" <*> parseDatePattern
               <|> WeekDayCst <$> parseWeekDay

parseMonthDayTerm :: Parser MonthDayTerm
parseMonthDayTerm = MonthDayCst <$> satM (\n -> 1 <= n && n <= 31) integer
                <|> MonthDayVar <$> parseVar  
                <|> MonthDayOf  <$ accept "day" <* accept "of " <*> parseDatePattern

parseMonthTerm :: Parser MonthTerm
parseMonthTerm = MonthOf <$ accept "month" <* accept "of" <*> parseDatePattern
             <|> MonthVar <$> parseVar
             <|> MonthCst . toEnum . (+ (-1)) <$> satM (\n -> 1 <= n && n <= 12) integer
             <|> MonthCst <$> parseMonth
             
parseYearTerm :: Parser YearTerm
parseYearTerm =   YearCst <$> parseYear
              <|> YearVar <$> parseVar
              <|> YearOf  <$ accept "year" <* accept "of" <*> parseDatePattern

parseDatePattern :: Parser DatePattern
parseDatePattern = parseAddDatePattern

parseAddDatePattern :: Parser DatePattern
parseAddDatePattern = do dt <- parseSimpleDatePattern  
                         ((DateDaysAfter dt <$ sChar '+' <*> integer
                            <|> DateDaysBefore dt <$ sChar '-' <*> integer) <* accept "days")
                            <|> return dt 

-- Tener en cuenta si hay que implementar modo ingles britanico o otros idioms
parseSimpleDatePattern :: Parser DatePattern
parseSimpleDatePattern =
            try ((\m d y -> DatePattern y m d) <$> parseMonthTerm
                                            <*> parseMonthDayTerm
                                            <*  sChar ','
                                            <*> parseYearTerm)
            <|> try ((\m d y -> DatePattern y m d)
                         <$> parseMonthTerm <* sChar '/'
                         <*> parseMonthDayTerm <* sChar '/'
                         <*> parseYearTerm)
            <|> try (Easter <$ accept "Easter" <*> parseYearTerm)
            <|> DateVar <$> parseVar
            <|> try (do n <- integer <* accept "days"
                        (flip DateDaysAfter n  <$ accept "after" <*> parseDatePattern
                           <|> flip DateDaysBefore n <$ accept "before" <*> parseDatePattern))
            <|> try (do accept "first"
                        (flip DateFirstDay <$ accept "day" <* accept "of" <*> parseMonthTerm <*> parseYearTerm
                          <|> (\w m y -> DateMonthFirstWeekDay y m w) <$> parseWeekDayTerm <* accept "in"
                                                                      <*> parseMonthTerm
                                                                      <*> parseYearTerm))
            <|> do  accept "last"
                    (flip DateLastDay <$ accept "day" <* accept "of" <*> parseMonthTerm <*> parseYearTerm
                      <|> (\w m y -> DateMonthLastWeekDay y m w) <$> parseWeekDayTerm <* accept "in"
                                                                 <*> parseMonthTerm
                                                                 <*> parseYearTerm)   
            <|> do wd <- parseWeekDayTerm
                   (flip DateWeekDayAfter wd <$ accept "after" <*> parseDatePattern
                     <|> flip DateWeekDayBefore wd <$ accept "before" <*> parseDatePattern)
            <|> (\nth wd dt -> DateNthWeekDay dt wd nth) <$> parseOrdinal
                                                         <*> parseWeekDayTerm <* accept "after"
                                                         <*> parseDatePattern
            <|> do accept "every"                                         
                   (flip DateNYearsSince 1 <$ accept "year" <* accept "since" <*> parseDatePattern 
                     <|> flip DateNMonthsSince 1 <$ accept "month" <* accept "since" <*> parseDatePattern 
                     <|> flip DateNDaysSince 1 <$ accept "day" <* accept "since" <*> parseDatePattern
                     <|> do n <- integer
                            (flip DateNYearsSince n <$ accept "years" <* accept "since" <*> parseDatePattern 
                              <|> flip DateNMonthsSince n <$ accept "months" <* accept "since" <*> parseDatePattern 
                              <|> flip DateNDaysSince n <$ accept "days" <* accept "since" <*> parseDatePattern)
                     )
            <|> sChar '(' *> parseAddDatePattern <* sChar ')' 

accept :: T.Text -> Parser T.Text
accept p = myToken (T.pack <$> (string . T.unpack) p) 

myToken :: Parser a -> Parser a
-- Parsec
myToken p = spaces *> p <* spaces
-- *Parsec
{-
-- Attoparsec
myToken p = skipSpace *> p <* skipSpace
-}

integer :: (Read a,Integral a) => Parser a
-- Parsec
integer = read <$> myToken (many1 digit)
-- *Parsec
{-
-- Attoparsec
integer = myToken decimal
-}

sChar :: Char -> Parser Char
sChar = myToken . char

parseHour :: Parser TimeOfDay
parseHour = do h <- parseH
               char ':'
               m <- parseM
               nm <- normOrMil
               verifyTime h m nm
          where  verifyTime h m Nothing = return $ TimeOfDay h m 0
                 verifyTime h m (Just h') = if h <= 12
                                            then return $ TimeOfDay ((h `mod` 12)+h') m 0
                                            else fail "Hour must be between 0 and 11"
                 parseH = satM (\h -> 0 <= h && h < 24) integer
                       <|> fail "The hour must be bettwen 0 and 23"
                 parseM = satM (\m -> 0 <= m && m < 60) integer
                        <|> fail "The minutes must be bettwen 0 and 59"
                 normOrMil =   return (Just 0) <* ( accept "am" <|>  accept "a.m." <|>  accept "AM" <|>  accept "A.M.")
                           <|> return (Just 12) <* ( accept "pm" <|>  accept "p.m." <|>  accept "PM" <|>  accept "P.M.")
                           <|> return Nothing

parseInterval :: Parser IntervalTime
parseInterval = IntervalTime <$> parseHour <* accept "-" <*> parseHour <*> parseDescription

parseList :: T.Text -> Parser a -> T.Text -> T.Text -> Parser [a]
parseList bg p sep end = accept bg *> parseListTerms p sep <* accept  end
                       where parseListTerms p sep = (:) <$> p <*> many (accept sep *> p)

parseListInt :: (Enum a) => T.Text -> Parser a -> T.Text -> T.Text -> Parser [a]
parseListInt bg p sep end = accept bg *> parseListTerms p sep <* accept  end
                       where parseListTerms p sep = do e <- p
                                                       do accept ".."
                                                          e' <- p
                                                          return [e .. e']
                                                         <|> (e:) <$> many (accept sep *> p)

optParseDaySchedule :: Parser DaySchedule
optParseDaySchedule = lookAhead (accept "[") *> parseDaySchedule
                    <|> return []

parseDaySchedule :: Parser DaySchedule
parseDaySchedule = do ds <- parseList "[" parseInterval ";" "]"
                      if wellDefinedSchedule ds
                      then return ds
                      else fail $ "Not well defined interval: "++show ds

parseWeekDayList :: Parser [WeekDay]
parseWeekDayList = nub <$> parseListInt "{" parseWeekDay "," "}"

parseInfoType :: Parser InfoType
parseInfoType =  parseEnum

parseWorkableType :: Parser (WorkableType,Bool,Description)
parseWorkableType = (NonBusiness,,) <$ accept "Non-business" <*> parseApplyHoliday <*> parseDescription 
                  <|> (\h des sc -> (Special sc,h,des)) <$ accept "Special" <* accept "schedule"
                                                    <*> parseApplyHoliday
                                                    <*> parseDescription <*> parseDaySchedule
                  <|> (\h des sc -> (Business sc,h,des))  <$ accept "Business" <* accept "schedule"
                                                      <*> parseApplyHoliday
                                                      <*> parseDescription <*> parseDaySchedule

parseDateVariable :: Parser DatePattern
parseDateVariable = accept "var" *> parseDateDef 

parseApplyHoliday :: Parser Bool
parseApplyHoliday = True <$ accept "for" <* accept "Holiday"
                  <|> return False

parseCalendarRule :: Parser CalendarRule
parseCalendarRule = do (ruleT,desc) <- ((,) <$> (InfoRule <$> parseInfoType) <*> parseDescription)
                                            <|> (\(x,h,d) -> (WorkableRule x h,d)) <$> parseWorkableType 
                            -- Could pass dayT argument
                       (DayRule ruleT desc <$> parseDatePatternDefList)
                              <|> ((\wds from to -> RangeRule ruleT desc from to wds)
                                   <$> parseWeekDayList
                                   <* accept "from" <*> parseDatePattern
                                   <* accept "to" <*> parseDatePattern)

parseCalendarTerm :: Parser ()
parseCalendarTerm = do r <- parseCalendarRule
                       updateState (\env -> env { rules = r : rules env })
                   <|> parseDateVariable *> return ()


parseEnv :: Parser RulesEnv
parseEnv = do many parseCalendarTerm
              getState
              
parseOnly:: Parser a -> T.Text -> Either ParseError a
parseOnly p s = runParser p (RulesEnv [] []) "unknown" s 

parseFile :: FilePath -> IO (Either ParseError RulesEnv)
parseFile path = parseOnly (parseEnv) <$> T.readFile path


{-
parseOnly (parseSimpleDatePattern) $ T.pack "January 12, 1994"
parseOnly (parseSimpleDatePattern) $ T.pack "12/15/1994"
parseOnly (parseSimpleDatePattern) $ T.pack "* 15, *"
parseOnly (parseSimpleDatePattern) $ T.pack "*/15/*"
parseOnly (parseSimpleDatePattern) $ T.pack "easter *"
parseOnly (parseSimpleDatePattern) $ T.pack "Monday after January 3, 1994"
parseOnly (parseSimpleDatePattern) $ T.pack "Monday after January *, *"
parseOnly (parseSimpleDatePattern) $ T.pack "10 days after easter 1994"
parseOnly (parseSimpleDatePattern) $ T.pack "first day of January *"
parseOnly (parseSimpleDatePattern) $ T.pack "first Monday in January *"
parseOnly (parseSimpleDatePattern) $ T.pack "last day of * *"
parseOnly (parseSimpleDatePattern) $ T.pack "last Monday in January *"
parseOnly (parseSimpleDatePattern) $ T.pack "third Monday after 01/1/1994"
parseOnly (parseSimpleDatePattern) $ T.pack "4th Monday after 01/1/1994"
parseOnly (parseSimpleDatePattern) $ T.pack "every year since 1/1/1994"
parseOnly (parseSimpleDatePattern) $ T.pack "every 4 years since 1/1/1994"
parseOnly (parseSimpleDatePattern) $ T.pack "every month since 1/1/1994"
parseOnly (parseSimpleDatePattern) $ T.pack "every 2 months since 1/1/1994"
parseOnly (parseSimpleDatePattern) $ T.pack "every day since 1/1/1994"
parseOnly (parseSimpleDatePattern) $ T.pack "every 30 days since 1/1/1994"
parseOnly (parseAddDatePattern) $ T.pack "January 12, 1994 - 3 days"

parseOnly (parseCalendarRule) $ T.pack "Holiday January 1, * 'Maria mother of God'"

parseOnly (parseRangeRule) $ T.pack "schedule since January 1, 2014 to January 20, 2014 {Monday .. Friday} [8:00 - 12:00pm]" 

parseOnly (parseRangeTerm) $ T.pack "rest period since January 1, * to January 20, *"
parseOnly (parseCalendarRule) $ T.pack "schedule since January 20, * to June 20, 2014 {Monday .. Friday} [8:00 - 12:00pm] 'First period'" 

parseOnly (parseCalendarRule) $ T.pack "Holiday every 4 years since January 20, 1937 'Inaguration Day'" 
parseOnly (parseCalendarRule) $ T.pack "Holiday every 4 years since 1/20/1937 'inaguration day'"
-}

{-
Examples MatchRule:

matchRules  <$> (either (const $ error "Error") id <$> (parseFile "Rules/churchRules.txt")) <*> return (2015,7,19)
smatchRules  <$> (either (const $ error "Error") id <$> (parseFile "Rules/churchRules.txt")) <*> return (2015,7,19)

-}        
